pennylane/devices/default_qutrit.py

```diff
@@ -0,0 +1,258 @@
+            return self._apply_ops[operation.base_name](state, axes, inverse=operation.inverse)
+
+        matrix = self._asarray(self._get_unitary_matrix(operation), dtype=self.C_DTYPE)
+
+        return self._apply_unitary(state, matrix, wires)
+
+    def _get_unitary_matrix(self, unitary):  # pylint: disable=no-self-use
+        """Return the matrix representing a unitary operation.
+
+        Args:
+            unitary (~.Operation): a PennyLane unitary operation
+
+        Returns:
+            array[complex]: Returns a 2D matrix representation of
+            the unitary in the computational basis.
+        """
+        return unitary.matrix()
+
+    @classmethod
+    def capabilities(cls):
+        capabilities = super().capabilities().copy()
+        capabilities.update(
+            model="qutrit",
+            supports_reversible_diff=True,
+            supports_inverse_operations=True,
+            supports_analytic_computation=True,
+            returns_state=True,
+        )
+        return capabilities
+
+    def _create_basis_state(self, index):
+        """Return a computational basis state over all wires.
+
+        Args:
+            index (int): integer representing the computational basis state
+
+        Returns:
+            array[complex]: complex array of shape ``[3]*self.num_wires``
+            representing the statevector of the basis state
+        """
+        state = np.zeros(3**self.num_wires, dtype=np.complex128)
+        state[index] = 1
+        state = self._asarray(state, dtype=self.C_DTYPE)
+        return self._reshape(state, [3] * self.num_wires)
+
+    @property
+    def state(self):
+        return self._flatten(self._pre_rotated_state)
+
+    def density_matrix(self, wires):
+        """Returns the reduced density matrix of a given set of wires.
+
+        Args:
+            wires (Wires): wires of the reduced system.
+
+        Returns:
+            array[complex]: complex tensor of shape ``(3 ** len(wires), 3 ** len(wires))``
+            representing the reduced density matrix.
+        """
+        dim = self.num_wires
+        state = self._pre_rotated_state
+
+        # Return the full density matrix by using numpy tensor product
+        if wires == self.wires:
+            density_matrix = self._tensordot(state, self._conj(state), axes=0)
+            density_matrix = self._reshape(density_matrix, (3 ** len(wires), 3 ** len(wires)))
+            return density_matrix
+
+        complete_system = list(range(0, dim))
+        traced_system = [x for x in complete_system if x not in wires.labels]
+
+        # Return the reduced density matrix by using numpy tensor product
+        density_matrix = self._tensordot(
+            state, self._conj(state), axes=(traced_system, traced_system)
+        )
+        density_matrix = self._reshape(density_matrix, (3 ** len(wires), 3 ** len(wires)))
+
+        return density_matrix
+
+    def _apply_unitary(self, state, mat, wires):
```

Comment:
Do you think it's worth implementing `_apply_unitary_einsum` as well (also the diagonal one)? The qubit version says it is faster for single- and two-qubit gates. I'm not sure how this was benchmarked but that might mean it is faster for single-qutrit operations as well, which would speed up some of the examples you've been running.

Reply:
I definitely agree. My aim is to create a PR once all major qutrit functionality is complete to add optimizations such as using `einsum` for small operations instead of `tensordot`, optimizing for diagonal operations, etc. I want to avoid adding it now to not make this PR too big.